redox/client.go

```diff
@@ -0,0 +1,258 @@
+		client.privateKey, err = jwt.ParseRSAPrivateKeyFromPEM([]byte(config.PrivateKeyPem))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return client, nil
+}
+
+func (c *client) GetSource() (source struct {
+	ID   *string `json:"ID"`
+	Name *string `json:"Name"`
+}) {
+	source.ID = &c.config.SourceId
+	source.Name = &c.config.SourceName
+	return
+}
+
+func (c *client) Send(ctx context.Context, payload interface{}) error {
+	req, err := c.getRequestWithFreshToken(ctx)
+	if err != nil {
+		return err
+	}
+
+	httpErr := &ErrorResponse{}
+	resp, err := req.
+		SetBody(payload).
+		SetHeader("Content-Type", "application/json").
+		SetError(httpErr).
+		Post("https://api.redoxengine.com/endpoint")
+
+	if err != nil {
+		return fmt.Errorf("error sending payload to redox: %w", err)
+	}
+	if resp.IsError() {
+		return fmt.Errorf("received %s error response when sending payload to redox: %w", resp.Status(), httpErr)
+	}
+
+	return nil
+}
+
+func (c *client) IsUploadFileEnabled() bool {
+	return c.config.UploadFileEnabled
+}
+
+func (c *client) UploadFile(ctx context.Context, fileName string, reader io.Reader) (*UploadResult, error) {
+	req, err := c.getRequestWithFreshToken(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if fileName == "" {
+		return nil, fmt.Errorf("file name is required")
+	}
+
+	httpErr := &ErrorResponse{}
+	uploadResult := &UploadResult{}
+	resp, err := req.
+		SetFileReader("file", fileName, reader).
+		SetResult(uploadResult).
+		SetError(httpErr).
+		Post("https://blob.redoxengine.com/upload")
+
+	if err != nil {
+		return nil, fmt.Errorf("error uploading redox: %w", err)
+	}
+	if resp.IsError() {
+		return nil, fmt.Errorf("received %s error response when uploading to redox: %w", resp.Status(), httpErr)
+	}
+
+	return uploadResult, nil
+}
+
+func (c *client) getRequest(ctx context.Context) *resty.Request {
+	return c.restyClient.R().SetContext(ctx)
+}
+
+func (c *client) getRequestWithFreshToken(ctx context.Context) (*resty.Request, error) {
+	if c.shouldRefreshToken() {
+		if err := c.obtainFreshToken(ctx); err != nil {
+			return nil, err
+		}
+	}
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.getRequest(ctx).SetAuthToken(c.token.AccessToken), nil
+}
+
+func (c *client) shouldRefreshToken() bool {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.token == nil || c.token.IsExpired(expirationDelta)
+}
+
+func (c *client) obtainFreshToken(ctx context.Context) error {
+	assertion, err := c.getSignedAssertion()
+	if err != nil {
+		return err
+	}
+
+	token := &Token{}
+	authErr := &AuthError{}
+
+	data := map[string]string{
+		"grant_type":            "client_credentials",
+		"client_assertion_type": "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
+		"client_assertion":      assertion,
+	}
+
+	c.logger.Debugw("obtaining a fresh token from redox")
+
+	resp, err := c.getRequest(ctx).
+		SetFormData(data).
+		SetResult(token).
+		SetError(authErr).
+		Post("https://api.redoxengine.com/v2/auth/token")
```

Comment:
Could the redox url prefix be extracted from inline as a const and reused anywhere else it gets used?

I notice they have different prefixes for blob/API so it'd be multiple const, but possibly worth extracting them for later.

Reply:
Good catch